GUI Jumps Fixes

champagne-cmd opened the pull request:
This PR fixes a bug that prevented `pintk` from working with jumps in .par files (#596). Jumps are accessed based on the number of jumps rather than the highest number within the jump flags, and jump flags are added for the .par file jumps. It also adds a sanity check test to TimingModel to check that maskParameters have associated TOAs (#708). However, this sanity check does not work with TNEQ parameters (possible issue or not?).

paulray commented:
This looks great!  Do you think it is ready to merge?
If so, I'll give it a quick test on some par files of mine and we can proceed.
Do you think you can make an issue for the scipy version problem? If it can be replicated by a short script outside of a notebook, that would make it easiest to troubleshoot.

champagne-cmd commented:
Yes it is ready to be tested and merge @paulray. I will make an issue for the scipy version.

paulray commented:
I'm testing this out now. I'll post comments for each issue I find. Some of them have nothing to do with this PR and are just pre-existing bugs in pint. First, I selected a few TOAs and hit “p” which is supposed to print out info about the TOAs. If fails like this:
```
   TOA              MJD Pre-fit residual (us) Pre-fit residual (phase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/paulr/env/pint/lib/python3.7/site-packages/matplotlib/cbook/__init__.py", line 196, in process
    func(*args, **kwargs)
  File "/Users/paulr/src/PINT/src/pint/pintk/plk.py", line 1204, in canvasKeyEvent
    self.print_info()
  File "/Users/paulr/src/PINT/src/pint/pintk/plk.py", line 867, in print_info
    inds = self.psr.selected_toas.table["index"][selected]
  File "astropy/table/_column_mixins.pyx", line 74, in astropy.table._column_mixins._ColumnGetitemShim.__getitem__
  File "astropy/table/_column_mixins.pyx", line 57, in astropy.table._column_mixins.base_getitem
  File "astropy/table/_column_mixins.pyx", line 69, in astropy.table._column_mixins.column_getitem
IndexError: boolean index did not match indexed array along dimension 0; dimension is 3 but corresponding boolean dimension is 62
```

paulray commented:
Next, I went to the profiles dir and tried `pintk NGC6440E.par NGC6440E.tim`. I highlighted the first 4 TOAs and hit "j". This added a JUMP component successfully, but when I click Fit, I get this warning:
```
WARNING: toas being fit must not all be jumped. Remove or uncheck at least one jump in the selected toas before fitting. [pint.pintk.plk]
```
If I proceed by turning off fitting for that JUMP, I get this:
```
INFO: JUMP1 set to 0 [pint.pintk.plk]
Pre-Fit Chi2:		59.574703 us^2
Pre-Fit Weighted RMS:	4.4562579 us
Post-Fit Chi2:		923249.44 us^2
Post-Fit Weighted RMS:	554.75203 us
          Parameter                   Pre-Fit	                Post-Fit	     Uncertainty        Difference          Diff/Unc
------------------------------------------------------------------------------------------------------------------------------------
     RAJ  hourangle         17:48:52.80034717	       17:48:52.90668557	       0.1890068     2.9538444e-05        0.21839885
    DECJ        deg        -20:21:29.38320870	      -20:21:14.00637420	       28.795901      0.0042713429        0.13001458
      DM   pc / cm3                224.113798	   224.11561956682609387	     0.003239902      0.0018215668       0.052135631
      F0         Hz       61.4854765543727595	   61.485476554903237493	   9.3569132e-10     5.3047799e-10         29.330701
      F1     Hz / s       -1.181337028639e-15	-1.189206254480437612e-15	   1.2833972e-17    -7.8692258e-18        -5.4577095
Exception in Tkinter callback
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/tkinter/__init__.py", line 1705, in __call__
    return self.func(*args)
  File "/Users/paulr/src/PINT/src/pint/pintk/plk.py", line 377, in fit
    self.fit_callback()
  File "/Users/paulr/src/PINT/src/pint/pintk/plk.py", line 560, in fit
    self.psr.fit(self.selected)
  File "/Users/paulr/src/PINT/src/pint/pintk/pulsar.py", line 512, in fit
    f.toas.select(no_jumps)
  File "/Users/paulr/src/PINT/src/pint/toa.py", line 978, in select
    self.table = self.table[selectarray].group_by("obs")
  File "/Users/paulr/env/pint/lib/python3.7/site-packages/astropy/table/table.py", line 3174, in group_by
    return groups.table_group_by(self, keys)
  File "/Users/paulr/env/pint/lib/python3.7/site-packages/astropy/table/groups.py", line 18, in table_group_by
    return _table_group_by(table, keys)
  File "/Users/paulr/env/pint/lib/python3.7/site-packages/astropy/table/groups.py", line 105, in _table_group_by
    out_keys = table_keys[indices[:-1]]
  File "/Users/paulr/env/pint/lib/python3.7/site-packages/astropy/table/table.py", line 1626, in __getitem__
    return self._new_from_slice(item)
  File "/Users/paulr/env/pint/lib/python3.7/site-packages/astropy/table/table.py", line 1145, in _new_from_slice
    newcol = col[slice_]
  File "astropy/table/_column_mixins.pyx", line 74, in astropy.table._column_mixins._ColumnGetitemShim.__getitem__
  File "astropy/table/_column_mixins.pyx", line 57, in astropy.table._column_mixins.base_getitem
  File "astropy/table/_column_mixins.pyx", line 69, in astropy.table._column_mixins.column_getitem
IndexError: index 0 is out of bounds for axis 0 with size 0
```

champagne-cmd commented:
@paulray I do not get that last error. When I load `pintk NGC6440E.par NGC6440E.tim`, add the jump to the first four, then fit them, it completes the fit just fine. I'm not sure why our results would be different...

paulray commented:
Next, I went into the "examples" dir and tried this: `pintk B1855+09_NANOGrav_9yv1.gls.par B1855+09_NANOGrav_9yv1.tim`

I highlighted a block of a few TOAs in the middle. I clicked "j" to add a JUMP, then did a Fit.  One issue is that adding the jump did not seem to add a second JUMP parameter to the GUI.  Then the fit didn't go because of the sanity check.:
```
/Users/paulr/src/PINT/src/pint/pintk/pulsar.py:308: RuntimeWarning: All-NaN axis encountered
  if np.isnan(np.nanmax(jump_nums)):
WARNING: The selected toa(s) overlap an existing jump. Remove all interfering jumps before attempting to jump these toas. [pint.pintk.pulsar]
ERROR: None [pint.pintk.plk]
INFO: Random Models set to 0 [pint.pintk.plk]
Pre-Fit Chi2:		106114.37 us^2
Pre-Fit Weighted RMS:	33.598754 us
Exception in Tkinter callback
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/tkinter/__init__.py", line 1705, in __call__
    return self.func(*args)
  File "/Users/paulr/src/PINT/src/pint/pintk/plk.py", line 377, in fit
    self.fit_callback()
  File "/Users/paulr/src/PINT/src/pint/pintk/plk.py", line 560, in fit
    self.psr.fit(self.selected)
  File "/Users/paulr/src/PINT/src/pint/pintk/pulsar.py", line 484, in fit
    fitter.fit_toas(maxiter=1)
  File "/Users/paulr/src/PINT/src/pint/fitter.py", line 653, in fit_toas
    self.model_init.maskPar_has_toas_check(self.toas)
  File "/Users/paulr/src/PINT/src/pint/models/timing_model.py", line 1360, in maskPar_has_toas_check
    "The maskParameter '%s' has no TOAs selected. " % maskpar
AttributeError: The maskParameter 'EFAC1' has no TOAs selected. 
```

This suggests a couple of fixes: (1) the sanity check does not need to do the check if parameter.frozen is True and (2) instead of printing just "EFAC1" is it possible to print the flag value for that parameter (which I assume is "-f L-wide_PUPPI ")

champagne-cmd commented:
Also, I am not getting any errors when I select TOAs and press 'p'.

champagne-cmd commented:
For the `pintk B1855+09_NANOGrav_9yv1.gls.par B1855+09_NANOGrav_9yv1.tim`, did you get a warning saying `The selected toa(s) overlap an existing jump. Remove all interfering jumps before attempting to jump these toas. [pint.pintk.pulsar]`? This was a functionality already present in `pintk`, it won't add jumps if they overlap. I can change this if this is incorrect. 
I see what you mean about the sanity check, and I will see about printing out the flag.

paulray commented:
> @paulray I do not get that last error. When I load `pintk NGC6440E.par NGC6440E.tim`, add the jump to the first four, then fit them, it completes the fit just fine. I'm not sure why our results would be different...

I don't either.  I just double checked that I'm on your branch:
```
% git lg |head
*   8eab663 - (HEAD -> champagne-cmd-GUI_mod) Merge branch 'GUI_mod' of https://github.com/champagne-cmd/PINT into champagne-cmd-GUI_mod (37 minutes ago) [Paul Ray]
|\  
| * 4c38daf - uncommented sanity check w/ new name (6 hours ago) [Chloe Champagne]
| * 3be2a93 - Scipy fix for Travis-CI (25 hours ago) [Chloe Champagne]
| * 064edc6 - added back removal of PhaseJump from fitted model if no JUMP (25 hours ago) [Chloe Champagne]
| * 8fffa3c - test to see if check causes travis failure (27 hours ago) [Chloe Champagne]
| * 4f48213 - fixed potential bug when deleting jumps (27 hours ago) [Chloe Champagne]
| * a18c162 - exempted TNEQ parameters from sanity check (2 days ago) [Chloe Champagne]
| * d6a8802 - sanity check added to TimingModel (2 days ago) [Chloe Champagne]
| * 05aa337 - bug fix - stopped adding extra toas to pintk jumps (2 days ago) [Chloe Champagne]
```
You don't get the warning at all?

paulray commented:
> Also, I am not getting any errors when I select TOAs and press 'p'.

Really?  That is so strange! Could it be an astropy version thing?  Mine is astropy 4.0

champagne-cmd commented:
> > @paulray I do not get that last error. When I load `pintk NGC6440E.par NGC6440E.tim`, add the jump to the first four, then fit them, it completes the fit just fine. I'm not sure why our results would be different...
> 
> I don't either. I just double checked that I'm on your branch:
> 
> ```
> % git lg |head
> *   8eab663 - (HEAD -> champagne-cmd-GUI_mod) Merge branch 'GUI_mod' of https://github.com/champagne-cmd/PINT into champagne-cmd-GUI_mod (37 minutes ago) [Paul Ray]
> |\  
> | * 4c38daf - uncommented sanity check w/ new name (6 hours ago) [Chloe Champagne]
> | * 3be2a93 - Scipy fix for Travis-CI (25 hours ago) [Chloe Champagne]
> | * 064edc6 - added back removal of PhaseJump from fitted model if no JUMP (25 hours ago) [Chloe Champagne]
> | * 8fffa3c - test to see if check causes travis failure (27 hours ago) [Chloe Champagne]
> | * 4f48213 - fixed potential bug when deleting jumps (27 hours ago) [Chloe Champagne]
> | * a18c162 - exempted TNEQ parameters from sanity check (2 days ago) [Chloe Champagne]
> | * d6a8802 - sanity check added to TimingModel (2 days ago) [Chloe Champagne]
> | * 05aa337 - bug fix - stopped adding extra toas to pintk jumps (2 days ago) [Chloe Champagne]
> ```
> 
> You don't get the warning at all?

No. I select the first 4 toas, jump them, fit them, and the only warning I get is about no ephemeris provided.

champagne-cmd commented:
These are the versions I have:
Python version: 3.6.9
SciPy version: 1.4.1, AstroPy version: 4.0, NumPy version: 1.18.4

paulray commented:
> > > @paulray I do not get that last error. When I load `pintk NGC6440E.par NGC6440E.tim`, add the jump to the first four, then fit them, it completes the fit just fine. I'm not sure why our results would be different...
> > 
> > You don't get the warning at all?
> 
> No. I select the first 4 toas, jump them, fit them, and the only warning I get is about no ephemeris provided.

Something funny is happening. I was a bit sloppy and said 4 when I meant 3. I just selected 4 TOAs, which was the first group of 3 plus one from the next block and now it works! I get a fit!
```
          Parameter                   Pre-Fit	                Post-Fit	     Uncertainty        Difference          Diff/Unc
------------------------------------------------------------------------------------------------------------------------------------
     RAJ  hourangle         17:48:52.80034691	       17:48:52.80034042	   0.00013603475    -1.8019507e-09    -1.3323442e-05
    DECJ        deg        -20:21:29.38333617	      -20:21:29.39049271	     0.036599315    -1.9879266e-06    -6.0510323e-05
      DM   pc / cm3     224.11379747853867105	   224.11432691146323798	     0.034959354     0.00052943292       0.015153073
      F0         Hz      61.48547655437250052	   61.485476554361973043	   2.9835957e-11    -1.0527478e-11       -0.58207614
      F1     Hz / s  -1.1813316387451346642e-15	-1.1806696681219663932e-15	   2.0749245e-18     6.6197062e-19         0.4591107
   JUMP1          s                       0.0	  -8.867543550434912e-06	   1.9987716e-05    -8.8675436e-06  
```

champagne-cmd commented:
> > > > @paulray I do not get that last error. When I load `pintk NGC6440E.par NGC6440E.tim`, add the jump to the first four, then fit them, it completes the fit just fine. I'm not sure why our results would be different...
> > > 
> > > 
> > > You don't get the warning at all?
> > 
> > 
> > No. I select the first 4 toas, jump them, fit them, and the only warning I get is about no ephemeris provided.
> 
> Something funny is happening. I was a bit sloppy and said 4 when I meant 3. I just selected 4 TOAs, which was the first group of 3 plus one from the next block and now it works! I get a fit!
> 
> ```
>           Parameter                   Pre-Fit	                Post-Fit	     Uncertainty        Difference          Diff/Unc
> ------------------------------------------------------------------------------------------------------------------------------------
>      RAJ  hourangle         17:48:52.80034691	       17:48:52.80034042	   0.00013603475    -1.8019507e-09    -1.3323442e-05
>     DECJ        deg        -20:21:29.38333617	      -20:21:29.39049271	     0.036599315    -1.9879266e-06    -6.0510323e-05
>       DM   pc / cm3     224.11379747853867105	   224.11432691146323798	     0.034959354     0.00052943292       0.015153073
>       F0         Hz      61.48547655437250052	   61.485476554361973043	   2.9835957e-11    -1.0527478e-11       -0.58207614
>       F1     Hz / s  -1.1813316387451346642e-15	-1.1806696681219663932e-15	   2.0749245e-18     6.6197062e-19         0.4591107
>    JUMP1          s                       0.0	  -8.867543550434912e-06	   1.9987716e-05    -8.8675436e-06  
> ```
Even with only 3, I still get it to fit. Hmmm...

champagne-cmd commented:
Could it be that something upstream changed that I haven't pulled?

paulray commented:
> Could it be that something upstream changed that I haven't pulled?

That is possible, I guess. I pulled your PR into nanograv/master 
```
* |   3486bb7 - (origin/master, origin/HEAD, nanograv/master, master) Merge pull request #732 from bshapiroalbert/ftest_pint (6 days ago) [Paul Ray]
```

So you could do `git merge nanograv/master` to merge that into your PR. (after doing `git fetch nanograv`)

champagne-cmd commented:
Ok, I merged the upstream and still don't have any trouble with getting the toa info by pressing 'p'. Jumps still seem to work when I first open `pintk`. However, I've noticed that if I click the `Reset` button and try to jump the exact same toas, or any other toa, I get this error message from my check: 
```
Traceback (most recent call last):
  File "/usr/lib/python3.6/tkinter/__init__.py", line 1705, in __call__
    return self.func(*args)
  File "/home/chloe/NRL_2020/PINT_personal/src/pint/pintk/plk.py", line 377, in fit
    self.fit_callback()
  File "/home/chloe/NRL_2020/PINT_personal/src/pint/pintk/plk.py", line 560, in fit
    self.psr.fit(self.selected)
  File "/home/chloe/NRL_2020/PINT_personal/src/pint/pintk/pulsar.py", line 484, in fit
    fitter.fit_toas(maxiter=1)
  File "/home/chloe/NRL_2020/PINT_personal/src/pint/fitter.py", line 653, in fit_toas
    self.model_init.maskPar_has_toas_check(self.toas)
  File "/home/chloe/NRL_2020/PINT_personal/src/pint/models/timing_model.py", line 1360, in maskPar_has_toas_check
    "The maskParameter '%s' has no TOAs selected. " % maskpar
AttributeError: The maskParameter 'JUMP1' has no TOAs selected. 
```
This is obviously problematic, and I need to go back and fix that. However, I am able to at least fit it in a new `pintk` session, whereas you seem to not be able to fit on even the first try @paulray.

champagne-cmd commented:
Also, I noticed that if I accidently click and drag in `pintk` without selecting any toas, I get this error message: 
```
Traceback (most recent call last):
  File "/home/chloe/.virtualenvs/nrl/lib/python3.6/site-packages/matplotlib/cbook/__init__.py", line 196, in process
    func(*args, **kwargs)
  File "/home/chloe/NRL_2020/PINT_personal/src/pint/pintk/plk.py", line 1021, in canvasReleaseEvent
    self.clickAndDrag(event)
  File "/home/chloe/NRL_2020/PINT_personal/src/pint/pintk/plk.py", line 1068, in clickAndDrag
    self.psr.selected_toas.select(self.selected)
  File "/home/chloe/NRL_2020/PINT_personal/src/pint/toa.py", line 978, in select
    self.table = self.table[selectarray].group_by("obs")
  File "/home/chloe/.virtualenvs/nrl/lib/python3.6/site-packages/astropy/table/table.py", line 3174, in group_by
    return groups.table_group_by(self, keys)
  File "/home/chloe/.virtualenvs/nrl/lib/python3.6/site-packages/astropy/table/groups.py", line 18, in table_group_by
    return _table_group_by(table, keys)
  File "/home/chloe/.virtualenvs/nrl/lib/python3.6/site-packages/astropy/table/groups.py", line 105, in _table_group_by
    out_keys = table_keys[indices[:-1]]
  File "/home/chloe/.virtualenvs/nrl/lib/python3.6/site-packages/astropy/table/table.py", line 1626, in __getitem__
    return self._new_from_slice(item)
  File "/home/chloe/.virtualenvs/nrl/lib/python3.6/site-packages/astropy/table/table.py", line 1145, in _new_from_slice
    newcol = col[slice_]
  File "astropy/table/_column_mixins.pyx", line 74, in astropy.table._column_mixins._ColumnGetitemShim.__getitem__
  File "astropy/table/_column_mixins.pyx", line 57, in astropy.table._column_mixins.base_getitem
  File "astropy/table/_column_mixins.pyx", line 69, in astropy.table._column_mixins.column_getitem
IndexError: index 0 is out of bounds for axis 0 with size 0
```
which seems to be similar to the warning you got when you first tried to fit the toas in `pintk NGC6440E.par NGC6440E.tim` @paulray. Maybe you accidently click and drag and it tried to fit no toas? The click and drag also seems to be buggy. Once you click and drag, it doesn't seem to want to stop if you didn't hit any toas.

champagne-cmd commented:
Summary of what's done and what bugs remain:

- I fixed the sanity check so that frozen parameters are bypassed. However, when you select a certain range of toas to fit in `pintk B1855+09_NANOGrav_9yv1.gls.par B1855+09_NANOGrav_9yv1.tim`, there is some sort of error in the fitter. It could be that all the selected toas are jumped from the par file and the fit function misses this. Fitting all the toas does fine for me.

- I have yet to figure out what is happening to the jumps after I hit the `Reset` button on the `tkinter` window. It seems that jumps added through `pintk` malfunction after hitting the `Reset` button and cannot be fitted, although they seem to be added correctly. Adding and fitting jumps works fine for me in a fresh `pintk` session. I will continue looking into this.

- I've cleaned up the `replicate` function in `timing_model.py` and added a unit test for it. The previous version seemed to add either a shallow copy or a deep copy of the components. This one just does a deep copy since `add_component` fully links the component to the current timing model and validates it (meaning some components would throw an error if added without certain parameters).

paulray commented:
OK, I found the line that seems to be wrong and is causing the crash when I hit "p":
https://github.com/nanograv/PINT/blob/3486bb7ede296008d868e727d3859c33e335a1cd/src/pint/pintk/plk.py#L867

See how it uses `selected_toas` but then also tries to index by `[selected]`.  This fails, but if I remove the final `[selected]`, then it works fine.

I'm not sure how this could ever have worked?  Also, debugging these GUI scripts are hard since I don't know how to get it to drop into the debugger so I can look at variables. I just had to keep adding print statements. :-)

paulray commented:
Also, I think my other fitting problems were user error. I didn't realize that when I selected TOAs that only those TOAs were used in the fit. When I select some TOAs, then hit "j" to add the JUMP, then hit "u" to unselect the TOAs and then fit, it does work.

paulray commented:
> * I have yet to figure out what is happening to the jumps after I hit the `Reset` button on the `tkinter` window. It seems that jumps added through `pintk` malfunction after hitting the `Reset` button and cannot be fitted, although they seem to be added correctly. Adding and fitting jumps works fine for me in a fresh `pintk` session. I will continue looking into this.

Maybe Reset goes back to a saved version of the initial model that has no JUMP component in it?

I also found that I get `AttributeError: No 'PhaseJump' in the timing model.` sometimes when trying to use "j" to remove the jump on a set of points.

champagne-cmd commented:
Yes @paulray that line is definitely not right in `plk.py`. And I have no idea why it works on my machine! 
I've mostly used print statements to debug, but you can also run `%run scripts/pintk.py ___ ___` in `ipython` and exit the `Tkinter` window whenever you want to check the variables with `%debug`.

champagne-cmd commented:
> > * I have yet to figure out what is happening to the jumps after I hit the `Reset` button on the `tkinter` window. It seems that jumps added through `pintk` malfunction after hitting the `Reset` button and cannot be fitted, although they seem to be added correctly. Adding and fitting jumps works fine for me in a fresh `pintk` session. I will continue looking into this.
> 
> Maybe Reset goes back to a saved version of the initial model that has no JUMP component in it?
> 
> I also found that I get `AttributeError: No 'PhaseJump' in the timing model.` sometimes when trying to use "j" to remove the jump on a set of points.

It does go back to the initial version, but when I add jumps again and try to fit, something weird is going on where it clears the jumps after adding them. The JUMP is there, but `select_toa_mask` and the flags get messed up.

I haven't gotten that `AttributeError` yet using "j" to delete them, I will see if I can reproduce it.

champagne-cmd commented:
@paulray when you ran `pintk B1855+09_NANOGrav_9yv1.gls.par B1855+09_NANOGrav_9yv1.tim` and selected the middle group of toas to fit and jump, you didn't intend to only jump those toas, correct? There is an error when you do this now with the fixed sanity check, from line 677 in `fitter.py`: `fac = M.std(axis=0)`. `fac` has some zero entries, which becomes problematic in later calculations. This seems unrelated to jumps, but is this an issue I should worry about?

paulray commented:
> @paulray when you ran `pintk B1855+09_NANOGrav_9yv1.gls.par B1855+09_NANOGrav_9yv1.tim` and selected the middle group of toas to fit and jump, you didn't intend to only jump those toas, correct? 

Yes indeed, I was just trying to add a JUMP that applied only to the selected TOAs. That isn't actually a very common use case I think, but I believe that is what the "j" key is supposed to be used for.

> There is an error when you do this now with the fixed sanity check, from line 677 in `fitter.py`: `fac = M.std(axis=0)`. `fac` has some zero entries, which becomes problematic in later calculations. This seems unrelated to jumps, but is this an issue I should worry about?

Hmm. That will take some more looking into. Looks like the design matrix has some rows with 0 standard deviation.

paulray commented:
As a separate note, in playing around and trying all the keypress options in pintk, I find that I get errors whenever I try to use "+", "-", ">", "<" and they don't seem to do what is intended.

champagne-cmd commented:
> Yes indeed, I was just trying to add a JUMP that applied only to the selected TOAs. That isn't actually a very common use case I think, but I believe that is what the "j" key is supposed to be used for.

Yes sorry, I misspoke, I meant fit; you didn't intend to fit only those toas. Jumping those toas produces a warning and aborts the operation because some of the toas selected were already jumped, and the application won't let you overlap existing jumps.

champagne-cmd commented:
> As a separate note, in playing around and trying all the keypress options in pintk, I find that I get errors whenever I try to use "+", "-", ">", "<" and they don't seem to do what is intended.

Should I open a separate PR to fix the keypress options, or bundle them into this PR? Also, what files were you using when pressing "j" wouldn't remove the jump? Were you attempting to remove jumps belonging to a par file?

paulray commented:
> > Yes indeed, I was just trying to add a JUMP that applied only to the selected TOAs. That isn't actually a very common use case I think, but I believe that is what the "j" key is supposed to be used for.
> 
> Yes sorry, I misspoke, I meant fit; you didn't intend to fit only those toas. Jumping those toas produces a warning and aborts the operation because some of the toas selected were already jumped, and the application won't let you overlap existing jumps.

Right. I was trying to add a JUMP to a selected set of TOAs, but then re-fit the full set of TOAs.

I think we should change the behavior.  Just because a few TOAs are selected should not mean that the fit is restricted to those TOAs. I think you should have to do something (perhaps click "z" for zoom, which aligns with the behavior of the tempo2 -gr plk plugin) before the TOAs included in the fit are restricted.

I think it is fine to keep adding to this PR for now.  Or, we can merge it and start a new one. Either is OK with me.

champagne-cmd commented:
@paulray I found the root cause of the `Reset` failures. The `selected_toas` object was only being updated when the user used "click and slide"; stationary clicks on individual points weren't updating the object, cause weird issues and inconsistencies. 

I think I will open another PR for general GUI fixes/updates since I also want to work on the "click and slide" feature.

As far as I've seen on my end, everything works now for this PR except for fitting a selected group of toas with jumps, which produces the zero standard deviation problem mentioned earlier.

paulray commented on src/pint/models/jump.py:
```diff
@@ -127,3 +131,10 @@ def print_par(self):
             result += jump_par.as_parfile_line()
         return result
+
+    def get_number_of_jumps(self):
```
@luojing1211 can you review the changes in this file and confirm that you agree  with them?

paulray commented on src/pint/models/jump.py:
```diff
@@ -127,3 +131,10 @@ def print_par(self):
+    def get_number_of_jumps(self):
+        return len(self.jumps)
+
+    def get_jump_param_objects(self):
```
@champagne-cmd can you add brief docstrings to these functions?

champagne-cmd replied:
Certainly!

paulray commented on src/pint/models/parameter.py:
```diff
@@ -1418,6 +1418,24 @@ def __repr__(self):
         return out
 
+    def create_new_index_copy(self, index):
+        """Sets up copy of current param with a new index."""
```
How about adding a bit more to this docstring about how this method is intended to be used?

paulray commented:
I think this is working great! I've been playing with it and I just have a couple requests before merging.
When I click "o" to print the model after fitting, I see the JUMP I added via the GUI.  
```
JUMP            -fe Rcvr_800    -0.0012262928755068423 1 9.669696128719508e-08
JUMP            jump 2    -8.120121154982946e-07 1 3.664338811262118e-07
```
That  line seems a bit odd. I guess it means that the flag controlling the jump is just "jump" and the value is "2" for that JUMP.  It might be more clear if the flag indicated that it was added by the GUI and if it looked like a normal TOA flag with the "-", so maybe more like:
```
JUMP  -gui_jump 2 -8.120121154982946e-07 1 3.664338811262118e-07
```
Is that difficult?  If I were to write out the TOAs at that point, would they get written with the -gui_jump flags on their TOA lines?  That would be cool.

paulray commented:
Also, I get these output lines when I run pintk:
```
INFO: This should show up [pint.pintk.plk]
INFO: This should also show up. test click revert, turn params on and off, and prefit model [pint.pintk.plk]
STARTING LOADING OF PULSAR J0740+6620.working.par
```
The first two should be made into `log.debug` instead of `log.info` and the third line should use `log.info` instead of `print()`. 
I know these preexisted your PR, but we might as well clean them up. :-)

champagne-cmd commented:
@paulray were these changes what you had in mind?

paulray commented:
Thanks!  We are very close now. When I add a jump in the GUI then fit and write out the par and tim files, the par file looks good with this line:
```
JUMP            -gui_jump 1     9.633478451331718e-06 1 1.0339745055818092e-05
```
But the .tim file looks like this:
```
pint 1949.609000 53478.2858714192189005 21.710 gbt  -format Princeton -ddm 0.0
pint 1949.609000 53483.2767051885165973 21.950 gbt  -format Princeton -ddm 0.0
pint 1949.609000 53489.4683897879295023 29.950 gbt  -format Princeton -ddm 0.0
pint 1724.609000 53679.8756459205251041 25.460 gbt  -format Princeton -ddm 0.0 -jump 1
pint 1874.609000 53679.8756453650567013 23.430 gbt  -format Princeton -ddm 0.0 -jump 1
pint 2024.609000 53679.8756449272914004 31.670 gbt  -format Princeton -ddm 0.0 -jump 1
pint 2174.609000 53679.8756445778781018 30.260 gbt  -format Princeton -ddm 0.0 -jump 1
pint 1949.609000 53679.8756451335503008 13.520 gbt  -format Princeton -ddm 0.0 -jump 1
```
so this won't quite work when I read it back in. The .tim file needs -gui_jump instead of -jump.

It looks like you modified parameter.py and timing_model.py to look for "jump" and change those to -gui_jump. Instead, is is possible to make the changes only in the GUI code and have the key for a GUI-added jump be "gui_jump" instead of "jump". Or is there some reason that is difficult?  Maybe they need to be called "jump" to deal with the bookkeeping of both pre-existing and GUI-added jumps?  It could be my idea was bad to call them "gui_jump".  It would also be fine to have the .tim file as above with "-jump 1", and just make the JUMP line in the par file be "-jump 1", but it does need the "-" in there to work.

champagne-cmd commented:
Ya, the "jump" flag in the toa table was what allowed the par file jumps and `pintk` jumps to mesh together. I could add another column in the toa table["flags"] for "gui_jumps", since that appears to be how the par file jumps keep track of their toas with their keys. That way, the key will be "-gui_jump" and show up that way in all the outputs, but `pintk` can still work with the "jump" flags implementation to unite par file and `pintk` added jumps. Does that sound good @paulray ?

paulray commented:
> Ya, the "jump" flag in the toa table was what allowed the par file jumps and `pintk` jumps to mesh together. I could add another column in the toa table["flags"] for "gui_jumps", since that appears to be how the par file jumps keep track of their toas with their keys. That way, the key will be "-gui_jump" and show up that way in all the outputs, but `pintk` can still work with the "jump" flags implementation to unite par file and `pintk` added jumps. Does that sound good @paulray ?

Yep, I think that sounds good!

champagne-cmd commented on src/pint/models/parameter.py:
```diff
@@ -1418,6 +1418,24 @@ def __repr__(self):
 
+    def create_new_index_copy(self, index):
+        """Sets up copy of current param with a new index."""
+        new_mask_param = maskParameter(
```
@luojing1211 the `new_param()` function does not copy over the `key` or `key_value` parameters. Should I simply add that to `new_param()`? I'm not sure if it would break something somewhere else but I can test it.

luojing1211 replied:
Let us fix it in an other PR.

luojing1211 replied:
Actually, why don't we fix it here? The key and key values should only be used in the mask parameter.   It should not break anything else.

champagne-cmd commented on src/pint/models/timing_model.py:
```diff
@@ -1350,6 +1336,18 @@ def as_parfile(
+            if "TNEQ" in str(par.name) or par.frozen:
+                continue
+            if par.is_mask and len(par.select_toa_mask(toas)) == 0:
+                raise AttributeError(
```
@luojing1211 Since this function is only meant for `maskParameters`, and since I need to use `getattr()` to find out if it's frozen, shouldn't I immediately narrow it down to `maskParameters` with `self.get_params_of_type_top("maskParameter")`? Otherwise, I would have to iterate through every single parameter in the model to see if it's frozen.

luojing1211 replied:
Get_params of type iterate through every single parameter as well. But it is a minor thing. We can fix it when we implementing a more robust sanity check.

paulray commented:
So, I just wrote out my .tim file after creating a JUMP in the GUI. They now look like this:
```
pint 1949.609000 53478.2858714192189005 21.710 gbt  -format Princeton -ddm 0.0
pint 1949.609000 53483.2767051885165973 21.950 gbt  -format Princeton -ddm 0.0
pint 1949.609000 53489.4683897879295023 29.950 gbt  -format Princeton -ddm 0.0
pint 1724.609000 53679.8756459205251041 25.460 gbt  -format Princeton -ddm 0.0 -gui_jump 1 -jump 1
pint 1874.609000 53679.8756453650567013 23.430 gbt  -format Princeton -ddm 0.0 -gui_jump 1 -jump 1
pint 2024.609000 53679.8756449272914004 31.670 gbt  -format Princeton -ddm 0.0 -gui_jump 1 -jump 1
pint 2174.609000 53679.8756445778781018 30.260 gbt  -format Princeton -ddm 0.0 -gui_jump 1 -jump 1
```
The par file has this line:
```
JUMP            -gui_jump 1      1.56140761202064e-05 1 1.2034463085648994e-05
```
So, this will work, but there is kind of a bonus `-jump 1` in the .tim file line.  Sorry if I have really confused things by asking for the "-gui_jump".

champagne-cmd commented:
> So, this will work, but there is kind of a bonus `-jump 1` in the .tim file line. Sorry if I have really confused things by asking for the "-gui_jump".
Yes, but I also thought it was good that now all toas that are jumped have the "-jump" flag, so now you can easily see where all the jumps are in the .tim file.

paulray commented:
Oh, I see, but for ones that come from the par file, they depend on the model. So, if I load a model that JUMPs some particular backend and a tim file then write out the tim file, I wouldn't want the TOA lines from that backend to get -jump flags, since if I use some other model it might not have the same backend JUMPed.  The manually added ones get their own special flag in the .tim file, which is great.  But, I think it would be better to not also have the -jump flag printed in the .tim file.

champagne-cmd commented:
> Oh, I see, but for ones that come from the par file, they depend on the model. So, if I load a model that JUMPs some particular backend and a tim file then write out the tim file, I wouldn't want the TOA lines from that backend to get -jump flags, since if I use some other model it might not have the same backend JUMPed. The manually added ones get their own special flag in the .tim file, which is great. But, I think it would be better to not also have the -jump flag printed in the .tim file.

Ah I see, I will fix that.

paulray commented:
OK, I'm ready to merge this.  OK @luojing1211 and @champagne-cmd ?

champagne-cmd commented:
That is good with me @paulray !

champagne-cmd commented:
Actually, are you seeing the conflict with `pulsar.py` @paulray ?

paulray commented:
Oops, I must have done that by merging my PR. Let me check.

luojing1211 commented:
I am good to merge it if the conflicts are fixed.

paulray commented:
I just fixed the conflict. It was from me removing the "set_pulse_nums" kwarg on Residuals().
